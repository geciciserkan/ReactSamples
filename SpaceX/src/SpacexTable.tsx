import React, { useState } from "react";
import {
  DataGrid,
  GridToolbarDensitySelector,
  GridToolbarFilterButton,
  GridColDef,
} from "@mui/x-data-grid";
import PropTypes from "prop-types";
import Box from "@mui/material/Box";
import ClearIcon from "@material-ui/icons/Clear";
import SearchIcon from "@material-ui/icons/Search";
import IconButton from "@mui/material/IconButton";
import TextField from "@mui/material/TextField";
import useAxios from "axios-hooks";

function escapeRegExp(value: any) {
  return value.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&");
}

function QuickSearchToolbar(props: any) {
  return (
    <Box
      sx={{
        p: 0.5,
        pb: 0,
        justifyContent: "space-between",
        display: "flex",
        alignItems: "flex-start",
        flexWrap: "wrap",
      }}
    >
      <div>
        <GridToolbarFilterButton />
        <GridToolbarDensitySelector />
      </div>
      <TextField
        variant="standard"
        value={props.value}
        onChange={props.onChange}
        placeholder="Searchâ€¦"
        InputProps={{
          startAdornment: <SearchIcon fontSize="small" />,
          endAdornment: (
            <IconButton
              title="Clear"
              aria-label="Clear"
              size="small"
              style={{ visibility: props.value ? "visible" : "hidden" }}
              onClick={props.clearSearch}
            >
              <ClearIcon fontSize="small" />
            </IconButton>
          ),
        }}
        sx={{
          width: {
            xs: 1,
            sm: "auto",
          },
          m: (theme) => theme.spacing(1, 0.5, 1.5),
          "& .MuiSvgIcon-root": {
            mr: 0.5,
          },
          "& .MuiInput-underline:before": {
            borderBottom: 1,
            borderColor: "divider",
          },
        }}
      />
    </Box>
  );
}
QuickSearchToolbar.propTypes = {
  clearSearch: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

interface Props {
  dataTestid?: string;
}

export const SpacexTable = ({ dataTestid = 'provision-data-test-id' }: Props) => {
  // get most recent 50 launches from SpaceX API.
  const [{ data, loading, error }] = useAxios(
    "https://api.spacexdata.com/v3/launches/past?sort=flight_number&order=desc&limit=50"
  );

  //parse the launch_date_local string into a date object.
  const parseDateString = (dateString: string) => {
    const date = new Date(dateString);
    return date.toLocaleDateString();
  };

  const tableRows = data?.map((item: any) => ({
    id: item.flight_number,
    launch_name: item.mission_name,
    launch_date_local: parseDateString(item.launch_date_local),
    rocket_details: item.rocket.rocket_name || "Launch details not available",
  }));
  const [searchText, setSearchText] = useState("");
  const [rows, setRows] = useState(tableRows);
  const [filteredRows, setFilteredRows] = useState(null);

  const requestSearch = (searchValue: React.SetStateAction<string>) => {
    setSearchText(searchValue);
    //filter the table rows based on the search text.
    const filteredRows = tableRows.filter((row: any) => {
      const regex = new RegExp(escapeRegExp(searchValue), "gi");
      return regex.test(row.launch_name);
    });
    setFilteredRows(filteredRows);
  };

  const columns: GridColDef[] = [
    { field: "id", headerName: "Flight ID" },
    { field: "launch_name", headerName: "Launch Name", width: 300 },
    { field: "launch_date_local", headerName: "Launch Date", width: 150 },
    { field: "rocket_details", headerName: "Rocket Details", width: 150 },
  ];
  return (
    <div data-testid={dataTestid}
      style={{
        height: 500,
        width: "40%",
      }}
    >
      <DataGrid   
        rows={filteredRows ? filteredRows : rows}
        columns={columns}
        components={{ Toolbar: QuickSearchToolbar }}
        componentsProps={{
          toolbar: {
            value: searchText,
            onChange: (event: {
              target: { value: React.SetStateAction<string> };
            }) => requestSearch(event.target.value),
            clearSearch: () => requestSearch(""),
          },
        }}
      />
    </div>
  );
}
