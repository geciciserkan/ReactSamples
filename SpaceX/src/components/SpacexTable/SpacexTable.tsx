import React, { useEffect, useState } from "react";
import {
  DataGrid,
  GridToolbarDensitySelector,
  GridToolbarFilterButton,
  GridColDef,
  GridRenderCellParams,
} from "@mui/x-data-grid";
import PropTypes from "prop-types";
import Box from "@mui/material/Box";
import ClearIcon from "@material-ui/icons/Clear";
import SearchIcon from "@material-ui/icons/Search";
import IconButton from "@mui/material/IconButton";
import TextField from "@mui/material/TextField";
import { useDispatch, useSelector } from "react-redux";
import {
  getLaunches,
  setSelectedLaunch,
} from "../../store/actions/launches.actions";
import { Button, CardMedia, Modal, Typography } from "@mui/material";
import "./styles.css";

function escapeRegExp(value: any) {
  return value.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, "\\$&");
}

function QuickSearchToolbar(props: any) {
  return (
    <Box
      sx={{
        p: 0.5,
        pb: 0,
        justifyContent: "space-between",
        display: "flex",
        alignItems: "flex-start",
        flexWrap: "wrap",
      }}
    >
      <div>
        <GridToolbarFilterButton />
        <GridToolbarDensitySelector />
      </div>
      <TextField
        variant="standard"
        value={props.value}
        onChange={props.onChange}
        placeholder="Searchâ€¦"
        InputProps={{
          startAdornment: <SearchIcon fontSize="small" />,
          endAdornment: (
            <IconButton
              title="Clear"
              aria-label="Clear"
              size="small"
              style={{ visibility: props.value ? "visible" : "hidden" }}
              onClick={props.clearSearch}
            >
              <ClearIcon fontSize="small" />
            </IconButton>
          ),
        }}
        sx={{
          width: {
            xs: 1,
            sm: "auto",
          },
          m: (theme) => theme.spacing(1, 0.5, 1.5),
          "& .MuiSvgIcon-root": {
            mr: 0.5,
          },
          "& .MuiInput-underline:before": {
            borderBottom: 1,
            borderColor: "divider",
          },
        }}
      />
    </Box>
  );
}
QuickSearchToolbar.propTypes = {
  clearSearch: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
};

interface Props {
  dataTestid?: string;
}

const SpacexTable = ({ dataTestid = "provision-data-test-id" }: Props) => {
  const [filteredRows, setFilteredRows] = useState([]);
  const [searchText, setSearchText] = useState("");

  const dispatch = useDispatch();
  const { isLoading, launches, errorMessage, selected } = useSelector(
    (state: any) => state.launches
  );

  useEffect(() => {
    dispatch(getLaunches());
  }, []);

  useEffect(() => {
    if (launches && launches.length > 0) {
      setFilteredRows(launches);
    }
  }, [launches]);

  const columns: GridColDef[] = [
    {
      field: "id",
      headerName: "Flight ID",
      sortable: false,
      filterable: false,
      disableColumnMenu: true,
    },
    {
      field: "launch_name",
      headerName: "Launch Name",
      width: 300,
      disableColumnMenu: true,
    },
    {
      field: "launch_date_local",
      headerName: "Launch Date",
      width: 150,
      type: "date",
      disableColumnMenu: true,
    },
    {
      field: "rocket_details",
      headerName: "Rocket Details",
      width: 150,
      sortable: false,
      filterable: false,
      disableColumnMenu: true,
      renderCell: (params: GridRenderCellParams) => {
        return (
          <Button
            onClick={() => dispatch(setSelectedLaunch(params.row.full_details))}
          >
            Click to View
          </Button>
        );
      },
    },
  ];

  const requestSearch = (searchValue: React.SetStateAction<string>) => {
    setSearchText(searchValue);
    const filteredBySearch = launches.filter((row: any) => {
      const regex = new RegExp(escapeRegExp(searchValue), "gi");
      return regex.test(row.launch_name);
    });
    setFilteredRows(filteredBySearch);
  };

  const handleCloseModal = () => {
    dispatch(setSelectedLaunch(null));
  };

  return (
    <div
      data-testid={dataTestid}
      className="ag-theme-alpine"
      style={{
        height: 500,
        width: "40%",
      }}
    >
      <DataGrid
        rows={filteredRows}
        columns={columns}
        components={{ Toolbar: QuickSearchToolbar }}
        componentsProps={{
          toolbar: {
            value: searchText,
            onChange: (event: {
              target: { value: React.SetStateAction<string> };
            }) => requestSearch(event.target.value),
            clearSearch: () => requestSearch(""),
          },
        }}
        loading={isLoading}
      />
      <Modal
        open={selected || false}
        onClose={handleCloseModal}
        aria-labelledby="modal-modal-title"
        aria-describedby="modal-modal-description"
      >
        <Box className="modal-container">
          <Typography id="modal-modal-title" variant="h5" component="h2">
            {selected?.mission_name}{" "}
            {selected?.launch_year && `(${selected?.launch_year})`}
          </Typography>
          <CardMedia
            component="img"
            image={selected?.links.mission_patch_small}
            alt={selected?.mission_name}
          />
          <Typography id="modal-modal-description" sx={{ mt: 2 }}>
            {selected?.details}
          </Typography>
        </Box>
      </Modal>
    </div>
  );
};

export default SpacexTable;
