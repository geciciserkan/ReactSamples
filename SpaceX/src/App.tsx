import React from "react";
import { Provider } from "react-redux";
import "./App.css";

import SpacexTable from "./components/SpacexTable/SpacexTable";
import configureStore from "./store/store";

const store = configureStore()

function App() {
  
  return (
    <Provider store={store}>

    <div className="App">
      <SpacexTable />
    </div>
    </Provider>
  );
}

export default App;
