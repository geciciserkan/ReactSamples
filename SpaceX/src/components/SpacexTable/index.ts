import SpacexTable from "./SpacexTable";

export default SpacexTable;