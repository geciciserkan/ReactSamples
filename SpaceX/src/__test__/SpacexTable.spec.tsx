import { render, screen, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import "@testing-library/dom";
import { SpacexTable } from "../SpacexTable";
import { unmountComponentAtNode } from "react-dom";
import { createMount } from "@material-ui/core/test-utils";
import "./App.spec";

let container: any = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("SpacexTable", () => {
  const dataTestid = "provision-data-test-id";
  it("should render the table", () => {
    // eslint-disable-next-line testing-library/no-unnecessary-act
    act(() => {
      render(<SpacexTable />, container);
    });
    expect(screen.getByTestId(dataTestid)).toBeInTheDocument();
  });
});

//Check the table columns and rows.
describe("Spacex Test", () => {
  const mount = createMount();

  test("Spacex Render", async () => {
    await act(async () => {
      mount(<SpacexTable />);
    });
    expect(await screen.findByRole("grid")).toBeInTheDocument();
    expect(await screen.findByRole("row")).toBeInTheDocument();
    expect(await screen.findAllByRole("columnheader")).toHaveLength(4);
  });
});

describe("Snapshot Testing", () => {
  const mount = createMount();

  it("Spacex Data Grid", () => {
    const wrapper = mount(<SpacexTable />);

    expect(wrapper).toMatchSnapshot();
  });
});
