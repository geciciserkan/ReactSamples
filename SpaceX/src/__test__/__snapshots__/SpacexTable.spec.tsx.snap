// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Spacex Data Grid 1`] = `ReactWrapper {}`;
